src/pdftoraw.c

```diff
@@ -0,0 +1,189 @@
+ * License along with this library. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include <unistd.h>
+#include <math.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <getopt.h>
+#include <gio/gio.h>
+#include <poppler.h>
+#include <gdk/gdk.h>
+
+#define DPI 150.0
+
+static struct option long_options[] = {
+    {"file",     required_argument, 0,  'f' },
+    {"pages",    no_argument,       0,  'p' },
+    {"test",     no_argument,       0,  't' },
+    {"width",    required_argument, 0,  'w' },
+    {"height",   required_argument, 0,  'W' },
+    {"raw",      required_argument, 0,  'r' },
+    {"help",     no_argument,       0,  'h' },
+    {0,          0,                 0,  0   }
+};
+
+static void
+print_usage(void) {
+    printf("Usage:\tpdftoraw  --filename=<filename>  --raw=<num-page>\n");
+    printf("\tpdftoraw  --filename=<filename>  --pages\n");
+    printf("\tpdftoraw  --filename=<filename>  --width=<num-page>\n");
+    printf("\tpdftoraw  --filename=<filename>  --height=<num-page>\n");
+    printf("\tpdftoraw  --help\n");
+}
+
+static int
+pdf_number_pages_get(PopplerDocument *doc)
+{
+    return poppler_document_get_n_pages(doc);
+}
+
+static unsigned char*
+pdf_page_raw_get(PopplerDocument *doc,
+                 int page_nr)
+{
+    cairo_surface_t *s = NULL;
+    cairo_t *cr = NULL;
+    unsigned char *data = NULL;
+    GdkPixbuf *pix = NULL;
+    double width = 0;
+    double height = 0;
+    int w = 0;
+    int h = 0;
+
+    PopplerPage *page = poppler_document_get_page(doc, page_nr);
+    poppler_page_get_size(page, &width, &height);
+    width = DPI * width / 72.0;
+    height = DPI * height / 72.0;
+    w = (int)ceil(width);
+    h = (int)ceil(height);
+    s = cairo_image_surface_create(CAIRO_FORMAT_ARGB32, w, h);
+    cr = cairo_create(s);
+    cairo_scale (cr, DPI/72.0, DPI/72.0);
+    cairo_save (cr);
+    poppler_page_render_for_printing(page, cr);
+    cairo_restore (cr);
+    pix = gdk_pixbuf_get_from_surface(s, 0, 0, w, h);
+    g_object_unref(page);
+    data = gdk_pixbuf_get_pixels (pix);
+    fwrite(data, 1, (w * h * 4), stdout);
+    return data;
+}
+
+static int
+pdf_page_width_get(PopplerDocument *doc,
+                   int page_nr)
+{
+    double width = 0;
+    int w = 0;
+
+    PopplerPage *page = poppler_document_get_page(doc, page_nr);
+    poppler_page_get_size(page, &width, NULL);
+    width = DPI * width / 72.0;
+    w = (int)ceil(width);
+    g_object_unref(page);
+    return w;
+}
+
+static int
+pdf_page_height_get(PopplerDocument *doc,
+                    int page_nr)
+{
+    double height = 0;
+    int h = 0;
+
+    PopplerPage *page = poppler_document_get_page(doc, page_nr);
+    poppler_page_get_size(page, NULL, &height);
+    height = DPI * height / 72.0;
+    h = (int)ceil(height);
+    g_object_unref(page);
+    return h;
+}
+
+int
+main(int argc, char **argv)
+{
+    PopplerDocument *doc = NULL;
+    GError *err = NULL;
+    GFile *in = NULL;
```

Comment:
`g_autoptr()` all of these.

Reply:
```
src/pdftoraw.c:150:15: error: unknown type name 'PopplerDocument_autoptr'
  150 |     g_autoptr(PopplerDocument) doc = NULL;
```

Reply:
For GFile, Fixed by 10e290f544

Reply:
Hmm. It is possible to define this with:

```c
G_DEFINE_AUTOPTR_CLEANUP_FUNC(PopplerDocument, g_object_unref);
```

Reply:
You still leak some poppler objects, so this is a good idea throughout I think.

Reply:
Fixed by 2c6c067